### PR TITLE
[AMDGPU] Check MIR after SIMemoryLegalizer instead of final ISA

### PR DESCRIPTION
**Pull request by jayfoad**
Since #72830 the memory legalizer tests have not shown s_waitcnt
instructions inserted by SIMemoryLegalizer because they have mostly been
removed by SIInsertWaitcnts.

Checking the MIR immediately after SIMemoryLegalizer runs fixes this so
you can see exactly what the pass has inserted.

### CODE REVIEW

File: llvm/test/CodeGen/AMDGPU/memory-legalizer-fence.ll [hunk 1]
```diff
@@ -1,2563 +1,3063 @@
...
-; GFX12-WGP:       ; %bb.0: ; %entry
-; GFX12-WGP-NEXT:    global_inv scope:SCOPE_SE
-; GFX12-WGP-NEXT:    s_endpgm
-;
-; GFX12-CU-LABEL: workgroup_acquire_fence:
-; GFX12-CU:       ; %bb.0: ; %entry
-; GFX12-CU-NEXT:    s_endpgm
+  ; GFX6-LABEL: name: workgroup_acquire_fence
+  ; GFX6: bb.0.entry:
+  ; GFX6-NEXT:   S_WAITCNT_soft 127
```
Comment (jayfoad) [inline llvm/test/CodeGen/AMDGPU/memory-legalizer-fence.ll:931]: Good example here of where all the waitcnts were removed.

**Comment by rampitec** [comment]
I am not a fan of the idea to convert all memory legalizer tests to mir. Maybe you can add some specific mir tests instead?

**Comment by rampitec** [comment]
> > I am not a fan of the idea to convert all memory legalizer tests to mir. Maybe you can add some specific mir tests instead?
> 
> Well all of `memory-legalizer-fence.ll` seems pretty useless at the moment since it doesn't show any of the waitcnts that were inserted. (OK it does still show cache invalidate instructions.)
> 
> As an alternative, how about changing SIInsertWaitcnts so that it does not remove redundant waitcnts at `-O0` and using that for all the memory legalizer tests?

I like this much more: we will see what is produced without decoding and it shall work faster too. Now these tests are very slow.

**Comment by jayfoad** [comment]
> > > I am not a fan of the idea to convert all memory legalizer tests to mir. Maybe you can add some specific mir tests instead?
> > 
> > 
> > Well all of `memory-legalizer-fence.ll` seems pretty useless at the moment since it doesn't show any of the waitcnts that were inserted. (OK it does still show cache invalidate instructions.)
> > As an alternative, how about changing SIInsertWaitcnts so that it does not remove redundant waitcnts at `-O0` and using that for all the memory legalizer tests?
> 
> I like this much more: we will see what is produced without decoding and it shall work faster too. Now these tests are very slow.

#90716

**Pull request closed without merging**